### ISSUE TITLE
Pattern match variable scope

### ISSUE THREAD

**Issue by Andrew-Hanlon** [dotnet/roslyn#15354]
In the [Roslyn Pattern Matching spec][1] it states that:

> The scope of a pattern variable is as follows:
> 
> If the pattern appears in the condition of an if statement, its scope
> is the condition and controlled statement of the if statement, but not
> its else clause.

However the latest Microsoft "What's new" [posts][2] and [presentations][3] are showing this example:

    public void PrintStars(object o)
    {
        if (o is null) return;     // constant pattern "null"
        if (!(o is int i)) return; // type pattern "int i"
        WriteLine(new string('*', i));
    }

> As you can see, the pattern variables – the variables introduced by a pattern – are similar to the out variables described earlier, in that they can be declared in the middle of an expression, and can be used within the nearest surrounding scope.

When / why has the scoping for if statements changed from the spec?
 
I feel that this scoping is fairly unintuitive and contradictory to all the other 'block' constructs like `catch` and `using`.

  [1]: https://github.com/dotnet/roslyn/blob/features/patterns/docs/features/patterns.md#scope-of-pattern-variables
  [2]: https://blogs.msdn.microsoft.com/dotnet/2016/08/24/whats-new-in-csharp-7-0/
  [3]: https://channel9.msdn.com/Events/Connect/2016/105

**Comment by DavidArno** [comment]
@Andrew-Hanlon,

It's long, but you can read all the gory details of why the language design team chose to "enhance" the language in this way at #12939.

TL;DR you aren't alone in thinking the change unintuitive and contradictory to the way scope have worked before. The team sadly don't care though and the change is here to stay.

**Comment by Andrew-Hanlon** [comment]
@DavidArno - Thanks for the information and source! 

I think this is a bad choice for the language, and I find it a shame that there is an assumptive justification around 'new keywords' being somehow worse that unintuitive and contradictory semantics.
